### PR TITLE
Add apiextensions v1 CustomResourceDefinition model

### PR DESCRIPTION
**Pull request by tg90nor**
- Support overriding field name
  &ensp;&ensp;Exposes the `Field.name` parameter so it can be different from the attribute name in the parent model.
- Add `SelfModel` class that can be used as `Field.type` to allow self-referencing models
  &ensp;&ensp;Field.type is then set to the parent model class during `Model.__init__`
- Add `JSONField` class that can represent any value that can be encoded/decoded by the python json module out of the box.
  &ensp;&ensp;Allows a field to contain any valid JSON value, but does no mapping to classes.
- Add stable apiextensions v1 CRD model
  &ensp;&ensp;Provides a complete mapping of `apiextensions.k8s.io/v1/customresourcedefinitions`

### Todo:
~Write tests~ ✔️

### CODE REVIEW

**Comment by tg90nor** [comment]
rebased

**Review by oyvindio — COMMENTED** [review]
I think the solution with the SelfModel type is good given the constraints of how fields and models work together. I don't really like that it is necessary to manipulate attributes of SelfModel fields outside the field class, but it seems to work, and I have a hard time seeing an alternative solution.
With the addition of SelfModel type and AnyField, I think it would be nice to extend the [field module tests](https://github.com/fiaas/k8s/blob/add-stable-v1-custom-resource-definition-model/tests/k8s/test_fields.py) with examples using these fields as well, including nested fields etc to try to catch any potential edge cases.
I've left a few other comments in line

File: k8s/base.py [hunk 1]
```diff
@@ -232,6 +234,9 @@ def __init__(self, new=True, **kwargs):
         for field in self._meta.fields:
             kwarg_names.discard(field.name)
             field.set(self, kwargs)
+            if field.type == SelfModel:
+                field.type = self.__class__
+                field.default_value_create_instance = False
```
[the remark this answers is not in the document]
Comment (tg90nor) [reply k8s/base.py:239]: I tried doing this in `__new__`  but the class won't be defined until `__new__` has returned, so it won't work. I'll look into whether this causes problems for `Field.set`.

File: k8s/models/apiextensions_v1.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+class WebhookClientConfig(Model):
+    caBundle = Field(six.text_type)
+    service = Field(ServiceReference)
+    url = Field(six.text_type)
+
+
+class WebhookConversion(Model):
+    clientConfig = Field(WebhookClientConfig)
+    conversionReviewVersions = ListField(six.text_type)
```
[the remark this answers is not in the document]
Comment (tg90nor) [reply k8s/models/apiextensions_v1.py]: Sure, I'll move them.

File: k8s/fields.py [hunk 3]
```diff
@@ -151,3 +152,22 @@ class RequiredField(Field):
     def is_valid(self, instance):
         value = self.__get__(instance)
         return value is not None and super(RequiredField, self).is_valid(instance)
+
+
+class AnyField(Field):
```
[the remark this answers is not in the document]
Comment (tg90nor) [reply k8s/fields.py:157]: AnyField is not suposed to contain lists or dicts of class instances, only built-in types that are (de)serializable out of the box. We could allow class instances in there, but then you'd get the weirdness of the class instances turning into dicts on `Model.save`. The alternative is to add some checks to prevent class instances from being used.

File: tests/k8s/test_fields.py [hunk 4]
```diff
@@ -39,6 +42,8 @@ class Meta:
     dict_field = Field(dict)
     _exec = Field(int)
     time_field = Field(datetime)
+    json_field = JSONField()
```
Comment (oyvindio) [inline tests/k8s/test_fields.py:45]: I think it could be useful to add a test that verifies the behavior when setting JSONField to a invalid value

**Review by oyvindio — COMMENTED** [review]
I think this looks mostly good, I've added a few comments/nitpicks in line

**Review by oyvindio — APPROVED** [review]
LGTM

**Review by oyvindio — APPROVED** [review]
